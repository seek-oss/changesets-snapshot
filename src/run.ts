import * as core from '@actions/core';
import { Package, getPackages } from '@manypkg/get-packages';

import { execWithOutput } from './utils';

type RunOptions = {
  script: string;
  cwd?: string;
};

type PublishedPackage = { name: string; version: string };

type PublishResult =
  | { published: true; publishedPackages: PublishedPackage[] }
  | { published: false };

export const run = async ({ script, cwd = process.cwd() }: RunOptions) => {
  const [runCommand, ...runArgs] = script.split(/\s+/);

  return execWithOutput(runCommand, runArgs, { cwd });
};

export const runPublish = async ({
  script,
  cwd = process.cwd(),
}: RunOptions): Promise<PublishResult> => {
  const prepublishScript = core.getInput('pre-publish');

  if (prepublishScript) {
    await execWithOutput(prepublishScript);
  }

  const changesetPublishOutput = await run({ script, cwd });

  const { packages, tool } = await getPackages(cwd);
  const releasedPackages: Package[] = [];

  if (tool !== 'root') {
    const newTagRegex = /New tag:\s+(@[^/]+\/[^@]+|[^/]+)@([^\s]+)/;
    const packagesByName = new Map(
      packages.map((x) => [x.packageJson.name, x]),
    );

    for (const line of changesetPublishOutput.stdout.split('\n')) {
      const match = line.match(newTagRegex);
      if (match === null) {
        continue;
      }
      const pkgName = match[1];
      const pkg = packagesByName.get(pkgName);
      if (pkg === undefined) {
        throw new Error(
          `Package "${pkgName}" not found.` +
            ' This is probably a bug in the action, please open an issue',
        );
      }
      releasedPackages.push(pkg);
    }
  } else {
    if (packages.length === 0) {
      throw new Error(
        'No package found.' +
          ' This is probably a bug in the action, please open an issue',
      );
    }
    const pkg = packages[0];
    const newTagRegex = /New tag:/;

    for (const line of changesetPublishOutput.stdout.split('\n')) {
      const match = line.match(newTagRegex);

      if (match) {
        releasedPackages.push(pkg);
        break;
      }
    }
  }

  if (releasedPackages.length) {
    return {
      published: true,
      publishedPackages: releasedPackages.map((pkg) => ({
        name: pkg.packageJson.name,
        version: pkg.packageJson.version,
      })),
    };
  }

  return { published: false };
};
