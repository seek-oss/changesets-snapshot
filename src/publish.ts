import * as core from '@actions/core';
import * as github from '@actions/github';
import { detect, getCommand } from '@antfu/ni';
import resolveFrom from 'resolve-from';

import { logger } from './logger';
import { ensureNpmrc, removeNpmrc } from './npm-utils';
import { run, runPublish } from './run';

const writeSummary = async ({
  title,
  message,
  codeBlock,
}: {
  title: string;
  message: string;
  codeBlock?: string;
}) => {
  core.summary.addHeading(title, 3);
  core.summary.addRaw(`<p>${message}</p>`, true);
  if (codeBlock) {
    core.summary.addCodeBlock(codeBlock);
  }
  await core.summary.write();
};

const failure = (reason: string) => {
  core.setFailed(reason);
  throw new Error(reason);
};

export const publishSnapshot = async () => {
  core.setOutput('published', false);
  core.setOutput('publishedPackages', []);

  const cwd = process.cwd();

  const githubToken = process.env.GITHUB_TOKEN;
  if (!githubToken) {
    throw failure('Unable to retrieve GitHub token');
  }

  const npmToken = process.env.NPM_TOKEN;
  if (!npmToken) {
    throw failure('Unable to retrieve NPM publish token');
  }

  const packageManager = await detect({ cwd });
  if (!packageManager) {
    throw failure('Unable to detect package manager');
  }

  const preVersionScript = core.getInput('pre-version');
  if (preVersionScript) {
    await run({ script: preVersionScript, cwd });
  }

  ensureNpmrc(npmToken);

  const branch = github.context.ref.replace('refs/heads/', '');
  const cleansedBranchName = branch.replace(/\//g, '-');
  const changesetsCli = resolveFrom(cwd, '@changesets/cli/bin.js');

  // Run the snapshot version
  const versionResult = await run({
    script: `node ${changesetsCli} version --snapshot ${cleansedBranchName}`,
    cwd,
  });

  if (versionResult.stderr.includes('No unreleased changesets found')) {
    logger.log(
      '\nNo changesets found. In order to publish a snapshot version, you must have at least one changeset committed.\n',
    );

    await writeSummary({
      title: '‚ö†Ô∏è No snapshot published',
      message:
        'No changesets found, skipping publish. If you want to publish a snapshot version, you may need to add a changeset for the relevant package.',
    });

    return;
  }

  const result = await runPublish({
    script: `node ${changesetsCli} publish --tag ${cleansedBranchName}`,
    cwd,
  });

  core.setOutput('published', result.published);

  if (result.published) {
    core.setOutput('publishedPackages', result.publishedPackages);

    const pkgNoun =
      result.publishedPackages.length === 1 ? 'snapshot' : 'snapshots';

    for (const { name, version } of result.publishedPackages) {
      await writeSummary({
        title: 'ü¶ã New snapshot published!',
        message: `Version: <code>${name}@${version}</code>`,
        codeBlock: getCommand(packageManager, 'add', [
          `${name}@${cleansedBranchName}`,
        ]),
      });
    }

    const newVersionsList = result.publishedPackages
      .map(({ name, version }) => `- ${name}@${version}`)
      .join('\n');

    logger.log(
      `
${result.publishedPackages.length} ${pkgNoun} published.
${newVersionsList}
`.trim(),
    );
  }

  removeNpmrc();
};
