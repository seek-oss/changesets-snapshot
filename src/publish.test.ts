import * as core from '@actions/core';
import * as github from '@actions/github';
import { detect } from '@antfu/ni';
import resolveFrom from 'resolve-from';

import { logger } from './logger';
import { publishSnapshot } from './publish';
import { run, runPublish } from './run';

jest.mock('@actions/github');
jest.mock('@actions/core');
jest.mock('@antfu/ni', () => ({
  ...jest.requireActual('@antfu/ni'),
  detect: jest.fn().mockName('@antfu/ni.detect'),
}));
jest.mock('resolve-from');
jest.mock('./npm-utils');
jest.mock('./run');
jest.mock('./logger');

const runMock = jest.mocked(run);
const runPublishMock = jest.mocked(runPublish);
const coreMock = jest.mocked(core);
const detectMock = jest.mocked(detect);

const getScriptCalls = <T extends jest.MockableFunction>(
  mockedFn: jest.MockedFn<T>,
) => mockedFn.mock.calls.map((args) => args[0].script);
const expectSummary = () => {
  expect(coreMock.summary.addHeading.mock.calls).toMatchSnapshot('summary');
  expect(coreMock.summary.addRaw.mock.calls).toMatchSnapshot('summary');
  expect(coreMock.summary.addCodeBlock.mock.calls).toMatchSnapshot('summary');
  expect(coreMock.summary.write).toHaveBeenCalled();
};

beforeEach(() => {
  process.env.GITHUB_TOKEN = '';
  process.env.NPM_TOKEN = '';
});

afterEach(() => {
  jest.clearAllMocks();
});

test.each(['yarn', 'npm', 'pnpm'])(
  'command output for %s',
  async (packageManager) => {
    process.env.GITHUB_TOKEN = '@github-token';
    process.env.NPM_TOKEN = '@npm-token';
    github.context.ref = 'feature/123-branch';
    runMock.mockResolvedValueOnce({
      code: 0,
      stdout: '',
      stderr: '',
    });
    runPublishMock.mockResolvedValueOnce({
      published: true,
      publishedPackages: [
        { name: '@multiple/package1', version: '1.2.3-SNAPSHOT' },
        { name: '@multiple/package-two', version: '1.2.3-SNAPSHOT' },
      ],
    });
    detectMock.mockResolvedValueOnce(
      packageManager as unknown as ReturnType<typeof detect>,
    );
    jest
      .mocked(resolveFrom)
      .mockImplementationOnce(
        (_fromDirectory, moduleId) => `/__mocked_node_modules__/${moduleId}`,
      );

    await publishSnapshot();

    expect(getScriptCalls(runMock)).toMatchSnapshot('run');
    expect(getScriptCalls(runPublishMock)).toMatchSnapshot('runPublish');
    expect(jest.mocked(logger).log.mock.calls[0]).toMatchSnapshot('logger.log');
    expectSummary();
  },
);

describe('error handling', () => {
  test('missing NPM token', async () => {
    process.env.GITHUB_TOKEN = '@github-token';

    await expect(() => publishSnapshot()).rejects.toMatchInlineSnapshot(
      `[Error: Unable to retrieve NPM publish token]`,
    );

    expect(coreMock.setFailed.mock.calls[0]).toMatchInlineSnapshot(`
      [
        "Unable to retrieve NPM publish token",
      ]
    `);
  });

  test('missing GitHub token', async () => {
    process.env.NPM_TOKEN = '@npm-token';

    await expect(() => publishSnapshot()).rejects.toMatchInlineSnapshot(
      `[Error: Unable to retrieve GitHub token]`,
    );

    expect(coreMock.setFailed.mock.calls[0]).toMatchInlineSnapshot(`
      [
        "Unable to retrieve GitHub token",
      ]
    `);
  });

  test('no changesets found', async () => {
    process.env.GITHUB_TOKEN = '@github-token';
    process.env.NPM_TOKEN = '@npm-token';
    runMock.mockResolvedValueOnce({
      code: 0,
      stdout: '',
      stderr: '\nNo unreleased changesets found\n',
    });
    detectMock.mockResolvedValueOnce('yarn');

    await publishSnapshot();

    expectSummary();
  });
});
