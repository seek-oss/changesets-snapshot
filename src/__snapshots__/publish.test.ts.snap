// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command output for npm: logger.log 1`] = `
[
  "2 snapshots published.
- @multiple/package1@1.2.3-SNAPSHOT
- @multiple/package-two@1.2.3-SNAPSHOT",
]
`;

exports[`command output for npm: run 1`] = `
[
  "node /__mocked_node_modules__/@changesets/cli/bin.js version --snapshot feature_123-branch",
]
`;

exports[`command output for npm: runPublish 1`] = `
[
  "node /__mocked_node_modules__/@changesets/cli/bin.js publish --tag feature_123-branch",
]
`;

exports[`command output for npm: summary 1`] = `
[
  [
    "ü¶ã New snapshot published!",
    3,
  ],
  [
    "ü¶ã New snapshot published!",
    3,
  ],
]
`;

exports[`command output for npm: summary 2`] = `
[
  [
    "<p>Version: <code>@multiple/package1@1.2.3-SNAPSHOT</code></p>",
    true,
  ],
  [
    "<p>Version: <code>@multiple/package-two@1.2.3-SNAPSHOT</code></p>",
    true,
  ],
]
`;

exports[`command output for npm: summary 3`] = `
[
  [
    "npm i @multiple/package1@feature_123-branch",
  ],
  [
    "npm i @multiple/package-two@feature_123-branch",
  ],
]
`;

exports[`command output for pnpm: logger.log 1`] = `
[
  "2 snapshots published.
- @multiple/package1@1.2.3-SNAPSHOT
- @multiple/package-two@1.2.3-SNAPSHOT",
]
`;

exports[`command output for pnpm: run 1`] = `
[
  "node /__mocked_node_modules__/@changesets/cli/bin.js version --snapshot feature_123-branch",
]
`;

exports[`command output for pnpm: runPublish 1`] = `
[
  "node /__mocked_node_modules__/@changesets/cli/bin.js publish --tag feature_123-branch",
]
`;

exports[`command output for pnpm: summary 1`] = `
[
  [
    "ü¶ã New snapshot published!",
    3,
  ],
  [
    "ü¶ã New snapshot published!",
    3,
  ],
]
`;

exports[`command output for pnpm: summary 2`] = `
[
  [
    "<p>Version: <code>@multiple/package1@1.2.3-SNAPSHOT</code></p>",
    true,
  ],
  [
    "<p>Version: <code>@multiple/package-two@1.2.3-SNAPSHOT</code></p>",
    true,
  ],
]
`;

exports[`command output for pnpm: summary 3`] = `
[
  [
    "pnpm add @multiple/package1@feature_123-branch",
  ],
  [
    "pnpm add @multiple/package-two@feature_123-branch",
  ],
]
`;

exports[`command output for yarn: logger.log 1`] = `
[
  "2 snapshots published.
- @multiple/package1@1.2.3-SNAPSHOT
- @multiple/package-two@1.2.3-SNAPSHOT",
]
`;

exports[`command output for yarn: run 1`] = `
[
  "node /__mocked_node_modules__/@changesets/cli/bin.js version --snapshot feature_123-branch",
]
`;

exports[`command output for yarn: runPublish 1`] = `
[
  "node /__mocked_node_modules__/@changesets/cli/bin.js publish --tag feature_123-branch",
]
`;

exports[`command output for yarn: summary 1`] = `
[
  [
    "ü¶ã New snapshot published!",
    3,
  ],
  [
    "ü¶ã New snapshot published!",
    3,
  ],
]
`;

exports[`command output for yarn: summary 2`] = `
[
  [
    "<p>Version: <code>@multiple/package1@1.2.3-SNAPSHOT</code></p>",
    true,
  ],
  [
    "<p>Version: <code>@multiple/package-two@1.2.3-SNAPSHOT</code></p>",
    true,
  ],
]
`;

exports[`command output for yarn: summary 3`] = `
[
  [
    "yarn add @multiple/package1@feature_123-branch",
  ],
  [
    "yarn add @multiple/package-two@feature_123-branch",
  ],
]
`;

exports[`error handling no changesets found: summary 1`] = `
[
  [
    "‚ö†Ô∏è No snapshot published",
    3,
  ],
]
`;

exports[`error handling no changesets found: summary 2`] = `
[
  [
    "<p>No changesets found, skipping publish. If you want to publish a snapshot version, you may need to add a changeset for the relevant package.</p>",
    true,
  ],
]
`;

exports[`error handling no changesets found: summary 3`] = `[]`;
